### DISCUSSION TITLE
Span-based Metrics

### DISCUSSION THREAD

**Discussion by Dhrumil-Sentry** [getsentry/sentry#81239] [Launch Feedback]
A few months ago we made the difficult decision to not release our generic Metrics solution which was in Beta. We understand this has been frustrating for many users, but after thorough evaluation, we determined there were strong reasons to take this step:

- **Lack of flexibility**: Traditional metrics solutions require developers to predefine tags and aggregation rules, which limits their ability to drill down and analyze data retrospectively. This approach hinders effective debugging, as unexpected issues cannot be investigated beyond predefined metrics. We believe that allowing developers to query raw data flexibly is the best way to resolve complex application health issues.
- **Low confidence in affordable pricing**: Our goal is to offer products at price points that make them accessible to all developers. However, we were not confident in finding a near-term solution that would have achieved this for the generic metrics solution we built.

## Our Path Forward

We’re excited to introduce Span Metrics, now available for a limited cohort of customers. It lets you easily aggregate and analyze span data without being restricted by predefined configurations or fixed parameters. Unlike traditional metrics setups, it gives you the flexibility to adjust your queries and groupings as needed.

### What Use Cases Can Span Metrics Solve?

Application Metrics supports a variety of use cases, offering improved visibility into application health and making it easier to debug and troubleshoot errors and slowdowns. 

1. **Track API Latency and Identify Anomalies:** Monitor the 95th percentile of time taken for API responses, grouped by customer, to identify if any users are overusing the system.
2. **Monitor Payload Sizes in Background Jobs**: Track the average payload size in background job processing to prevent slowdowns.
3. **Track Token Usage for Third-Party APIs**: Ensure that requests to external APIs like OpenAI stay within limits by tracking token usage per request which you slice by your customer segments
4. **Monitor LCP (Largest Contentful Paint) for Specific Users or Pages**: Monitor LCP for high-value users on the checkout page to identify page load issues that may impact revenue.

### Attach attributes/tags to your spans via SDKs

You can add any custom attribute to your spans via our SDKs and use these attributes for filtering and aggregation in our UX. Here’s an example from our [React SDK](https://docs.sentry.io/platforms/javascript/guides/react/tracing/instrumentation/custom-instrumentation/#adding-span-attributes):

```
const span = Sentry.getActiveSpan();
if (span) {
  span.setAttribute("customerType", "Premium");
  // Or set multiple attributes at once:
  span.setAttributes({
    "payloadSize": 42,
    "cached": true,
  });
}
```

### Use the new Trace Explorer to query and aggregate your spans

To support flexible, real-time queries on application metrics, we’re introducing an upgraded **Trace Explorer UX** (Go to Explore→Traces) making it easy to:

- Filter and aggregate span data based on attributes like duration, custom fields, and endpoints.
- Visualize high-level metrics or drill down into specific span samples.
- Switch seamlessly between viewing span aggregates and underlying trace details.

Example: You can query the p75 and p90 of LCP for your webpages and aggregate these score by custom attribute customerType

![image](https://github.com/user-attachments/assets/9bf251c3-5109-4da0-b692-e03105ce8265)

### Extrapolating for Sample rates

Since we’re aggregating spans to compute your aggregates, we will automatically extrapolate your aggregations like count, percentiles etc to account for your sampling rates.  This will help us show you reasonably accurate aggregates even if you apply client side sample rates

## Your feedback is appreciated

We’ve rolled this feature to a limited cohort of customers. Please reach out to [feedback-tracing@sentry.io](mailto:feedback-tracing@sentry.io) to get access

We’re actively working on this querying experience so please feel free to share any feedback 

- Is the filtering UX easy to use? We want to ensure you quickly find the values to filter by
- Do you need any advance features (equations, multi-queries etc)?

Your input will help us build a better experience, so please drop your comments and questions below. We’re excited to help you unlock the full potential of your span data with Sentry.

**Comment by Dhrumil-Sentry** [comment]
Hi All - We have released this ability to freely query and aggregate any span based metric to all Early Adopter Customers !! If you want access just go to your Org Settings and turn on the Early Adopter Flag. 

Enterprise customers can reach out to their account reps for more details

**Reply by munkii** [reply]
Are these span based metrics sent to Sentry even when there is no an Exception/Error being logged?

**Reply by Dhrumil-Sentry** [reply]
@munkii - Yes, these span based metrics rely on your tracing data being sent to Sentry and will be sent even when there is no Exception/Error

Here's a [blog](https://blog.sentry.io/find-and-fix-performance-bottlenecks-with-sentrys-trace-explorer/#setting-up-and-using-span-metrics-in-the-new-trace-explorer) that can help you get started 😄

**Reply by munkii** [reply]
Thye are queried\reported under the "Errors"  within "Discover" section right?

And we would send them using 

`SentrySdk.CaptureMessage(logEntry.Message, level: ToSentryLevel(logEntry.Level));`

**Reply by Dhrumil-Sentry** [reply]
@munkii -That's incorrect. You need to [send traces](https://docs.sentry.io/platforms/javascript/tracing/) to Sentry. You can then [send any custom spans or attach custom attributes](https://docs.sentry.io/platforms/javascript/tracing/instrumentation/custom-instrumentation/)

You see this data under Explore->Traces in the side navigation menu
<img width="1484" alt="image" src="https://github.com/user-attachments/assets/b0974703-9ce6-4eff-9ec5-24276eac9ea0" />

**Reply by kadet1090** [reply]
So if I understand correctly I won't have metrics about all users, only about sampled traces? 

This is a problem because for most cases I want my tracing sample rate to be quite low, but I would love to obtain metrics about non-sampled sessions too.

**Reply by Dhrumil-Sentry** [reply]
@kadet1090 - Every span in a sampled trace that gets sent to Sentry has a sampling rate attached to it. We use this to extrapolate what the true counts for your query is with reasonable accuracy.

We show such extrapolation subtext in the UX too
<img width="341" alt="image" src="https://github.com/user-attachments/assets/cc096d23-bc11-4a38-8a01-6f8d8ee0040c" />

If you ever write a query which has less accuracy due to low sampling rates we'll make that clear too. This allows you to compute any metric based on your tracing data with reasonable accuracy and you don't have to pay extra for metrics or cardinality.

Also - if you're on our [latest span based pricing plan](https://sentry.io/pricing/) then you can send a lot more tracing data to Sentry at an economic rate.

**Reply by munkii** [reply]
> @munkii -That's incorrect. You need to [send traces](https://docs.sentry.io/platforms/javascript/tracing/) to Sentry. You can then [send any custom spans or attach custom attributes](https://docs.sentry.io/platforms/javascript/tracing/instrumentation/custom-instrumentation/)
> 
I guessed it was incorrect but to be honest it is not all that clear how to send Trace information using the MAUI SDK. I have looked at the [Custom Instrumentation page for MAUI](https://docs.sentry.io/platforms/dotnet/guides/maui/tracing/instrumentation/custom-instrumentation/). 

That talks about 

```
var transaction = SentrySdk.StartTransaction(
  "test-transaction-name",
  "test-transaction-operation"
);

transaction.Finish()
```

We use that to trace some offline synchronisation work, that is long running and has a start and end so "Transactions" makes sense. However if we take one of our more interesting analaytic events that seems to not make less sense. When a user does a series of medical tests in our app we show them the full list and preselect what we consider to be the best choice. If they choose a different result we log that. Would that be logged as 

```
var transaction = SentrySdk.StartTransaction(
  "CaptureOverride",
  "User overrode recommended capture"
);

transaction.Finish()
```

**Reply by Dhrumil-Sentry** [reply]
@munkii -Sorry that our MAUI custom instrumentation docs are not fully up to date, We're fixing that.

But if you take a look at this [doc](https://docs.sentry.io/platforms/dotnet/guides/maui/tracing/instrumentation/custom-instrumentation/queues-module/) on how to instrument Queue spans in MAUI - You'll see examples for how to create custom spans and attach attributes to them

So in your case you could fetch the existing transaction and create child spans in case a user chose a different test

(@bruno-garcia please correct me if I am wrong, Bruno's our .Net expert 😄 )

**Comment by Dhrumil-Sentry** [comment]
Span Metrics are generally available to all our SaaS customers !

You can now run on-the-fly aggregations over billions of spans — grouped, filtered, and visualized by any span attribute, even custom ones you send - without worrying about cardinality

Why it matters:

Span Metrics unlocks fast answers to questions like:

- “What’s the p90 duration of image spans, grouped by server.address?”
- “What’s the average cart.total from checkout spans, grouped by user.region?”
- “Which DB queries are slowest on p50 across services?”
- “How many times did payment.status: retry happen in the last hour?”

No need to pre-define metrics or log extra data — if it's on the span, you can query it.

Check out our [blog](https://blog.sentry.io/announcing-trace-explorer-and-span-metrics/) to learn more
